### ISSUE TITLE
Proposal: weave only selected assemblies

### ISSUE THREAD

**Issue by mmnosek** [Starcounter/Home#198]
By default, weaver processes all dependencies. Everything that shouldn't be weaved needs to be explicitly written in the `weaver.ignore` file. In some projects (a lot of external libraries/assemblies) it is a pain. You need to exclude them one by one as not all of them are reported as problematic ones.  From my experience, often there is more that needs to be excluded than included.

Would it be possible to make default weaving configurable? If turned on, works as it is right now, if turned off - developer needs to specify what will be weaved explicitly (in the file similar to `weaver.ignore`).

It's just an idea, that would make some projects probably easier.

cc @joozek78 @per-samuelsson

**Comment by k-rus** [comment]
Will it destroy the first experience? Does it mean that a new developer, trying Starcounter, will need to add his/her the only C# file/project into weaver white list to execute the simplest application in Starcounter?

**Comment by mmnosek** [comment]
That's why I suggested to leave it by default as it is (all weaved). Optionally, it should be somehow possible to turn weaving by default off, in project configuration for example. It will help in the projects where there is a big codebase already and won't destroy first experience/backward compatibility.

**Comment by per-samuelsson** [comment]
> You need to exclude them one by one as not all of them are reported as problematic ones.

If they are not problematic, you don't need to exclude them. Exclusion was introduced for 2 reasons:

1. For assemblies that failed to load properly in the weaver, e.g. native assemblies, assemblies that was not solvable in the given context, and
2. Performance - if you know nothing in a certain assembly is relevant to the weaver (i.e. no database classes, and no references to such), you can speed up weaving some by excluding it

Regarding the proposal, it's doable of course, but I'm not sure it's a prio. As always, we have to judge. And withthe increased complexity it will introduce - more flexibility with optional opt-in / out-out means more things to explain, maintain, document - I'm not sure it's all good.

**Comment by mmnosek** [comment]
I've made a proposal after some bad experience with `weaver` and multiple tries to make it work.

> And withthe increased complexity it will introduce - more flexibility with optional opt-in / out-out means more things to explain, maintain, document - I'm not sure it's all good.

I agree it's not good. Ideal solution would be if exclusion is only for reason 2 (performance). It would be better to invest time into solving this issues with weaver. Do we have any significant plans to improve weaver?

**Comment by per-samuelsson** [comment]
> Do we have any significant plans to improve weaver?

Well, yeah, we already have an ongoing effort rewriting it from scratch, using Roslyn instead of PostSharp as the engine. However, we are still facing some uncertainties there, including what to be expected from C# team, so it's progressing slower than we would want.

As for the current, IL-based PostSharp-driven weaver, it's somewhat problematic because we are using a version of PostSharp that have features we depend on that are no longer part of PostSharp.

The hard part for me (and us) is to find the right balance - how much time to spend on trying to fix flaws in the current weaver versus working with the new one.

As it happens, I'm actually currently working on an issue that is somewhat related to this one: #166. Here's a recent comment that could be interesting for this case too: https://github.com/Starcounter/Home/issues/166#issuecomment-321229034.

All thoughts on this broad topic are highly appreciated!

**Comment by per-samuelsson** [comment]
> It would be better to invest time into solving this issues with weaver.

@mmnosek, relevant new findings that could interest you: https://github.com/Starcounter/Home/issues/166#issuecomment-321766616

**Comment by miyconst** [comment]
@per-samuelsson since the `weaver.ignore` is no longer needed, is this issue obsolete as well?

**Comment by per-samuelsson** [comment]
> @per-samuelsson since the `weaver.ignore` is no longer needed, is this issue obsolete as well?

The redesign of the weaver is in 2.4, so there, yes, I guess this issue is obsolete.

And as I wrote earlier, I don't think this is a prio, considering redesign we already implemented in 2.4 and  the fact that second generation weaver is already being developed. On top of that, we have no outstanding customer complaints AFAIK.

I vote we close it. @mmnosek WDYT?
